### ISSUE TITLE
[Feature] Resume offline runs

### ISSUE THREAD

**Issue by brenerrr** [wandb/wandb#2423]
Whenever I need to resume a run, I have to make sure it has been synced to the cloud before I try to resume it. It would be nice to be able to resume offline runs. 

The following code demonstrates the problem. If you are running on "dryrun" mode, you will not be able to resume the run. 

```python
import os
import wandb

os.environ["WANDB_MODE"] = "dryrun" # This does not work
# os.environ["WANDB_MODE"] = "run" # This works
wandb.init(project="test")
wandb.config.update({"first_run": True})
run_id = wandb.run.id
wandb.run.save()
wandb.finish()
# Resuming run
wandb.init(project="test", id=run_id, resume="must")
if wandb.run.resumed: print("Resuming run")
else: print("Run not resummed")
wandb.finish()
```

**Comment by tyomhak** [comment]
Hey @brenerrr , thank you for the feature request! We have this feature in our plans, but I can't give an ETA for it yet.

**Comment by abaisero** [comment]
Is there any update on this?  I have the exact same issue.

**Comment by aidanjdonohue** [comment]
Unfortunately no movement yet. I wouldn't expect something until client version 1.0 sometime next year.

**Comment by abaisero** [comment]
This would be a very welcome feature;  for anybody else who is experiencing the same problem, I'd like to add some info.  The original comment was implying that it's possible to resume offline runs if they have been synced in-between.  However, when I tried doing that, wandb would show the following warning:

> wandb: WARNING `resume` will be ignored since W&B syncing is set to `offline`.  Starting a new run with run id <run_id>.

Seems like it's saying that this would not work, and that a completely separate run was going to start.  However, I have verified that it does indeed work as desired, and that the resumed logs are associated with the same run, rather than a new run, i.e., the warning message is inaccurate.

TLDR:  syncing before resuming allows you to resume offline runs, despite what the warning message says.

**Comment by abaisero** [comment]
As a fairly important addendum to the above.  To resume offline runs correctly, also take into account issue https://github.com/wandb/client/issues/2971

**Comment by savvaki** [comment]
I can confirm that the resumed runs behave as @abaisero mentioned. Multiple runs with the same resume ID are stored in the wandb folder which all need to be synced for the runs to be merged on wandb.

**Comment by bg4xsd** [comment]
Test Case :

id = '20230121'
wandb.init(  id=id,  resume="auto",  mode="offline",  project="test-2023", config={"first_run": True}   )

Msg ：
 WARNING `resume` will be ignored since W&B syncing is set to `offline`. Starting a new run with run id  '20230121'

Result:
Using wandb sync ./wandb/offline-A, and wandb sync ./wandb/offline-B.
It seems B overwrites A, and can not get the A+B result.

**Comment by nate-wandb** [comment]
@bg4xsd this is the current behavior. If a run is started in offline mode and is passed a `run_id` you will get a new run. We currently don't have a way to get the A+B behavior you are looking for with offline runs but I have re-raised this ticket to get this looked at by the engineering team to see if this can be implemented.

Thank you,
Nate

**Comment by Chulabhaya** [comment]
Hi, is there any progress on this issue? It would be extremely useful to have!

**Comment by nate-wandb** [comment]
Hi @Chulabhaya, it has not been assigned to get implemented yet.  I'll make a new feature request for this so that it can be triaged again by the engineering team. I'll follow up once they get a chance to look into this. 

Thank you,
Nate

**Comment by abaisero** [comment]
An update on my comment from a couple years ago;  it seems like the resuming of offline runs has stopped working at some point.  Something weird is happening, where the **summary** of the resumed run is correct, but the UI won't show any of the resumed data.

Here is a minimal working example:

mwe.py:
```python
#!/usr/bin/env python
import pickle
import random
from dataclasses import dataclass

import wandb
import wandb.util


@dataclass
class ProgramState:
    """Program"""

    x: int = 0
    y: float = 0.0


def save_data(data):
    """saves generic data to file"""
    with open('checkpoint.pkl', 'wb') as f:
        pickle.dump(data, f)


def load_data():
    """loads generic data from file"""
    with open('checkpoint.pkl', 'rb') as f:
        return pickle.load(f)


def save_program_state(program_state: ProgramState):
    """saves the program state with metadata"""
    assert wandb.run is not None
    save_data(
        {
            'run_id': wandb.run.id,
            'program_state': program_state,
        }
    )


def load_program_state(wandb_kwargs) -> ProgramState:
    """loads the program state"""
    try:
        data = load_data()
    except FileNotFoundError:
        program_state = ProgramState()
    else:
        wandb_kwargs.update(
            {
                'resume': 'must',
                'id': data['run_id'],
            }
        )
        program_state = data['program_state']

    return program_state


def run_program(program_state: ProgramState) -> ProgramState:
    x_start, y = program_state.x, program_state.y
    x_stop = x_start + 100

    for x in range(x_start, x_stop):
        y += random.random() - 0.5
        wandb.log({'x': x, 'y': y}, step=x)  # note explicit step

    return ProgramState(x_stop, y)


def main():
    wandb_kwargs = {'tags': ['mwe']}
    program_state = load_program_state(wandb_kwargs)

    wandb.init(**wandb_kwargs)

    program_state = run_program(program_state)
    save_program_state(program_state)


if __name__ == '__main__':
    main()
```

mwe.sh:
```bash
#!/bin/bash

export WANDB_API_KEY=<hidden>
export WANDB_DIR=$PWD
export WANDB_ENTITY=<hidden>
export WANDB_PROJECT=mwe-checkpoint
export WANDB_MODE=offline

{
  python mwe.py 2>&1
  # wandb sync --sync-all --clean
  wandb sync --sync-all
} 2>&1 | tee -a mwe.log
```

These are the produced logs (note, the WARNING form the second run has always been present, even when this whole setup was working in the past):
```
wandb: Tracking run with wandb version 0.15.0
wandb: W&B syncing is set to `offline` in this directory.  
wandb: Run `wandb online` or set WANDB_MODE=online to enable cloud syncing.
wandb: Waiting for W&B process to finish... (success).
wandb: 
wandb: Run history:
wandb: x ▁▁▁▁▂▂▂▂▂▃▃▃▃▃▃▄▄▄▄▄▅▅▅▅▅▅▆▆▆▆▆▇▇▇▇▇▇███
wandb: y ▆▅▆▇▅▅▆▇▇▇▅▄▃▃▄█▅█▇▆▅▄▅▅▄▆▃▃▄▃▂▁▃▅▆▅█▆▆▆
wandb: 
wandb: Run summary:
wandb: x 99
wandb: y -0.54565
wandb: 
wandb: You can sync this run to the cloud by running:
wandb: wandb sync /home/bais/git/asym-rlpo/mwe-wandb-checkpoint/wandb/offline-run-20230426_132119-uplyialw
wandb: Find logs at: ./wandb/offline-run-20230426_132119-uplyialw/logs
Find logs at: /home/bais/git/asym-rlpo/mwe-wandb-checkpoint/wandb/debug-cli.bais.log
Syncing: https://wandb.ai/abaisero/mwe-checkpoint/runs/uplyialw ... done.
wandb: WARNING `resume` will be ignored since W&B syncing is set to `offline`. Starting a new run with run id uplyialw.
wandb: Tracking run with wandb version 0.15.0
wandb: W&B syncing is set to `offline` in this directory.  
wandb: Run `wandb online` or set WANDB_MODE=online to enable cloud syncing.
wandb: Waiting for W&B process to finish... (success).
wandb: 
wandb: Run history:
wandb: x ▁▁▁▁▂▂▂▂▂▃▃▃▃▃▃▄▄▄▄▄▅▅▅▅▅▅▆▆▆▆▆▇▇▇▇▇▇███
wandb: y ▁▁▁▂▂▂▃▃▃▁▂▂▄▅▅▆▆▅▇▆▅▆▆▅▄▆▇███▇▇▆█▇▆▇▆█▆
wandb: 
wandb: Run summary:
wandb: x 199
wandb: y 1.1967
wandb: 
wandb: You can sync this run to the cloud by running:
wandb: wandb sync /home/bais/git/asym-rlpo/mwe-wandb-checkpoint/wandb/offline-run-20230426_132258-uplyialw
wandb: Find logs at: ./wandb/offline-run-20230426_132258-uplyialw/logs
Find logs at: /home/bais/git/asym-rlpo/mwe-wandb-checkpoint/wandb/debug-cli.bais.log
Syncing: https://wandb.ai/abaisero/mwe-checkpoint/runs/uplyialw ... done.
```

I run mwe.sh a couple times.  Each time the MWE is run, 100 new logs are made.  Then, I check the UI and I see the following inconsistency.  Looking at the run summary, I notice that all 200 logs seem to be present:

![image](https://user-images.githubusercontent.com/1794938/234656967-08b6b531-1700-4ccb-a7ca-417f1f962c80.png)

but looking at the actual run plot, only the first 100 logs are shown:

![image](https://user-images.githubusercontent.com/1794938/234657118-3ef893ca-ec4f-4794-b7f0-3eb31a8b12e3.png)

It's pretty problematic for my workflow that resuming offline runs has stopped working;  it's pretty much a necessity in my case.  And from what I've been seeing in other threads, I'm far from the only one.  I'm hoping that this is still possible in some way, just like it was possible in the past, but I'm a bit afraid that simply the functionality has stopped working for good, rather than there being an actual way of getting this to work.  If that is indeed the case, I sadly I don't expect a solution to come out any day now, which leaves me a bit uncertain how to move on with my work.

**Comment by abaisero** [comment]
New update on the above.

I got in contact with the support team via email, and got the solution to issue I describe in my last reply.  The solution to getting resumed offline runs to work again, per the MWE shown above, is to add the `--append` option to the `wandb sync` command.  I've been stress-testing this setup and it works.

Good logging yall!

**Comment by LedaguenelArthur** [comment]
Hi everyone,

If I'm not wrong @abaisero 's solution (while very convenient in many cases ;-) ) does not allow to retrieve information about the previous run to execute the next one, which can be a crucial feature in some use cases (eg. retrieving the last learning rate of the scheduler).

Am I right on this ? Or is there a way to do so ?

Best regards,
Arthur

**Comment by nate-wandb** [comment]
Hi @LedaguenelArthur, that is correct. You could use the [Public Api ](https://docs.wandb.ai/guides/track/public-api-guide#export-data)though to get information out of previous run (config, summary metrics, etc...). I understand this is more of a workaround though and we are looking to fully support resuming offline runs in the future.

**Comment by FlyingAnt2018** [comment]
> Hey @brenerrr , thank you for the feature request! We have this feature in our plans, but I can't give an ETA for it yet.

Almost one year has passed. Is there any progress?

**Comment by anmolmann** [comment]
> > Hey @brenerrr , thank you for the feature request! We have this feature in our plans, but I can't give an ETA for it yet.
> 
> Almost one year has passed. Is there any progress?

Hey @FlyingAnt2018, this feature request is still pending - we're working on some other performance improvements related requests. 
However, we do currently have an alternative way to ensure that offline runs are synced before we can resume them. [For instance](https://docs.wandb.ai/ref/cli/wandb-sync#docusaurus_skipToContent_fallback): `wandb sync --clean-force` will delete only synced runs. You should be able to help you achieve the desired results via this flag `clean-force`. Please let me know if this flag is helpful for your use-case.

**Comment by streethagore** [comment]
Is there any update on this matter?
I would crucially need to resume offline runs.
My compute nodes on my cluster does not have internet access so I must run them in offline mode.

**Comment by SevKod** [comment]
> Is there any update on this matter? I would crucially need to resume offline runs. My compute nodes on my cluster does not have internet access so I must run them in offline mode.

Same here in the exact situation. Only 20 hours allowed and no internet access on my cluster 😟

**Comment by anmolmann** [comment]
Hey folks, apologies for the delay on this one. [Here's](https://github.com/wandb/wandb/issues/5865#issuecomment-2069690715) a quick update for you all.
>  our eng is actively working on this request - resuming a run from a checkpoint via the graphQL API. We'll keep you updated on this thread once this is merged and deployed.

**Comment by anmolmann** [comment]
Hey folks, this https://github.com/wandb/wandb/pull/7711 implements the SDK changes needed for "run rewinding," a long-requested feature that allows users to truncate the history logged to a run at an intermediate step and resume logging from that step. This will be part of our [next SDK release](https://github.com/wandb/wandb/releases).

**Comment by agarwal-abhinav** [comment]
I am a bit confused - I have tried to resume offline run using the `--append` method suggested above as well as the suggestions [here](https://github.com/wandb/wandb/issues/2423#issuecomment-1797130004). None of those work. The log and overview pages update but the charts don't. Setting any kind of a `resume` flag for an offline run simply results in a warning saying that `resume` will be ignored. 
Wandering how [this](https://github.com/wandb/wandb/issues/2423#issuecomment-2149926563) will solve the problem? Am I missing something?

**Comment by kabouzeid** [comment]
Any updates on this? This is really annoying...

**Comment by omerkolcak** [comment]
I was facing the same problem and --append flag or any other solution did not work for me as well. What it worked for me is that instead of resuming the training with using resume='must' and explicitly giving the same wandb id, I continue training from the model checkpoint and log the results independently from the previous run. After that using wand api, I concatanted the results to the new wandb log run. 

```
import wandb

ENTITY = "your-entity"
PROJECT = "your-project"
RUN_1 = '8gr2wqnv'      
RUN_2 = '7jx71d39'      

api = wandb.Api()

first_run = api.run(f"{ENTITY}/{PROJECT}/{RUN_1}")
first_df = first_run.history()

second_run = api.run(f"{ENTITY}/{PROJECT}/{RUN_2}")
second_df = second_run.history()

con_df = pd.concat([first_df, second_df]).reset_index(drop=True)

wandb.init(
    entity=ENTITY,
    project=PROJECT,
    id='concat_log',
)

for index, row in con_df.iterrows():
    metrics_dict = row.to_dict()

    step = int(metrics_dict['trainer/global_step'])

    wandb.log(metrics_dict, step=step)
```

Note that I am using pytorch lightning to train my model, therefore it handles the step numbers for training, they should match.

**Comment by bernardo-torres** [comment]
Any updates on this? I have been having this issue a lot recently
